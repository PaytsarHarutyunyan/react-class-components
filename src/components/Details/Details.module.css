.detailsContainer {
    width: 500px;
    height: 600px;
    position: absolute;
    top: 0;
    right: 0;
    z-index: 1000;
    background: #bababa;
    overflow: hidden; 
}

.detailsLink {
    position: absolute;
    top: 0;
    right: 20px;
    text-decoration: none;
    color: rgb(0, 0, 0);
    font-size: 20px;
}

.detailsContent {
    display: flex;
    gap: 10px;
    margin: 5px 0 12px 15px;
}

.detailsKey {
    font-size: 20px;
}

.detailsValue {
    font-size: 20px;
}

h2 {
    margin-left: 220px;
}