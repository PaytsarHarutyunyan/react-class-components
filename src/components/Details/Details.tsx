import React, { FC, useEffect, useState } from 'react';
import { useQueryParams } from '@/hooks/useQueryParams';
import { BASE_URL } from '@/constants/commonConstants';
import { getData } from '@/api';
import { ResultItem } from '@/types';
import { Link } from 'react-router-dom';
import styles from './Details.module.css';

type DetailsProps = {
    selectedTabPath: string;
};

const Details: FC<DetailsProps> = ({ selectedTabPath }) => {
    const [loading, setLoading] = useState(false);
    const [details, setDetails] = useState<ResultItem | null>(null);

    const { getItemId, getSelectedPage } = useQueryParams();
    const itemId = getItemId();

    useEffect(() => {
        if (!itemId) return undefined;
        setLoading(true);
        getData(`${BASE_URL}${selectedTabPath}\\${itemId}`).then((result) => {
            setLoading(false);
            setDetails(result);
        });
    }, [itemId, selectedTabPath]);

    if (loading) {
        return <span>Loading...</span>;
    }

    if (!itemId || !details) return null;

    return (
        <div
            className={styles.detailsContainer}
            onClick={(e) => {
                e.stopPropagation();
            }}
        >
            <Link
                className={styles.detailsLink}
                onClick={() => {
                    setDetails(null);
                }}
                to={`${location.pathname}?page=${getSelectedPage()}`}
            >
                X
            </Link>
            <div>
                <h2>Details</h2>
                {Object.entries(details).map(([key, value]) => (
                    <div className={styles.detailsContent} key={key}>
                        <h4 className={styles.detailsKey}>{key}:</h4>
                        <div className={styles.detailsValue}>{value}</div>
                    </div>
                ))}
            </div>
        </div>
    );
};

export default Details;
