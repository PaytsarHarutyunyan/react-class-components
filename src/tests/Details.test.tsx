import React from 'react';
import { screen, render } from '@testing-library/react';
import Details from '../components/Details/Details';

describe('Details Component', () => {
    const mockData = {
        name: 'Test Item',
        description: 'This is a test item description',
    };

    it('displays details', () => {
        render(<Details itemDetails={mockData} selectedPage={1} />);

        expect(screen.getByTestId('details-container')).toBeInTheDocument();
        expect(screen.getByText('Details')).toBeInTheDocument();
        expect(screen.getByText('name:')).toBeInTheDocument();
        expect(screen.getByText('Test Item')).toBeInTheDocument();
        expect(screen.getByText('description:')).toBeInTheDocument();
        expect(screen.getByText('This is a test item description')).toBeInTheDocument();
    });
});
