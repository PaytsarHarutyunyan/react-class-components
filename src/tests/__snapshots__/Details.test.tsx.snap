// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Component Make sure the detailed card component correctly displays the detailed card data 1`] = `
<div>
  <div
    class="detailsContainer"
  >
    <a
      class="detailsLink"
      href="/?page=1"
    >
      X
    </a>
    <div>
      <h2>
        Details
      </h2>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          name
          :
        </h4>
        <div
          class="detailsValue"
        >
          R2-D2
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          height
          :
        </h4>
        <div
          class="detailsValue"
        >
          96
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          mass
          :
        </h4>
        <div
          class="detailsValue"
        >
          32
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          hair_color
          :
        </h4>
        <div
          class="detailsValue"
        >
          n/a
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          skin_color
          :
        </h4>
        <div
          class="detailsValue"
        >
          white, blue
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          eye_color
          :
        </h4>
        <div
          class="detailsValue"
        >
          red
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          birth_year
          :
        </h4>
        <div
          class="detailsValue"
        >
          33BBY
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          gender
          :
        </h4>
        <div
          class="detailsValue"
        >
          n/a
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          homeworld
          :
        </h4>
        <div
          class="detailsValue"
        >
          https://swapi.dev/api/planets/8/
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          films
          :
        </h4>
        <div
          class="detailsValue"
        >
          https://swapi.dev/api/films/1/
          https://swapi.dev/api/films/2/
          https://swapi.dev/api/films/3/
          https://swapi.dev/api/films/4/
          https://swapi.dev/api/films/5/
          https://swapi.dev/api/films/6/
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          species
          :
        </h4>
        <div
          class="detailsValue"
        >
          https://swapi.dev/api/species/2/
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          vehicles
          :
        </h4>
        <div
          class="detailsValue"
        />
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          starships
          :
        </h4>
        <div
          class="detailsValue"
        />
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          created
          :
        </h4>
        <div
          class="detailsValue"
        >
          2014-12-10T15:11:50.376000Z
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          edited
          :
        </h4>
        <div
          class="detailsValue"
        >
          2014-12-20T21:17:50.311000Z
        </div>
      </div>
      <div
        class="detailsContent"
      >
        <h4
          class="detailsKey"
        >
          url
          :
        </h4>
        <div
          class="detailsValue"
        >
          https://swapi.dev/api/people/3/
        </div>
      </div>
    </div>
  </div>
</div>
`;
